File: mobile/ios/Runner/Core/NetworkMonitorApiImpl.swift
import Network

/// Streams NWPathMonitor updates to Flutter.
///
/// iOS has no public equivalent of Android's VALIDATED capability, so
/// `internetValidated` is nil while a route exists (the Dart side falls back
/// to its own reachability probe) and false when there is no route at all.
class NetworkMonitorApiImpl: NetworkMonitorApi {
  private let monitor = NWPathMonitor()
  private let queue = DispatchQueue(label: "NetworkMonitorApi")
  private let events: NetworkMonitorEvents
  private var observing = false
  private var currentPath: NWPath?

  init(events: NetworkMonitorEvents) {
    self.events = events
    monitor.pathUpdateHandler = { [weak self] path in
      guard let self else { return }
      self.currentPath = path
      guard self.observing else { return }
      let status = Self.status(from: path)
      DispatchQueue.main.async {
        self.events.onStatusChanged(status: status) { _ in }
      }
    }
    monitor.start(queue: queue)
    currentPath = monitor.currentPath
  }

  deinit {
    monitor.cancel()
  }

  func getCurrentStatus() throws -> NativeNetworkStatus {
    return Self.status(from: currentPath)
  }

  func startObserving() throws {
    observing = true
  }

  func stopObserving() throws {
    observing = false
  }

  private static func status(from path: NWPath?) -> NativeNetworkStatus {
    guard let path, path.status == .satisfied else {
      return NativeNetworkStatus(
        hasTransport: false,
        transports: [],
        internetValidated: false,
        isExpensive: false
      )
    }

    var transports: [NativeTransportType] = []
    if path.usesInterfaceType(.wifi) {
      transports.append(.wifi)
    }
    if path.usesInterfaceType(.cellular) {
      transports.append(.cellular)
    }
    if path.usesInterfaceType(.wiredEthernet) {
      transports.append(.ethernet)
    }
    if path.usesInterfaceType(.other) {
      transports.append(.vpn)
    }
    if transports.isEmpty {
      transports.append(.other)
    }

    return NativeNetworkStatus(
      hasTransport: true,
      transports: transports,
      internetValidated: nil,
      isExpensive: path.isExpensive
    )
  }
}

File: mobile/android/app/src/main/kotlin/com/seagate/curator/stxphotos/android/networkmonitor/NetworkMonitorApiImpl.kt
package com.seagate.curator.stxphotos.android.networkmonitor

import android.content.Context
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities
import android.os.Handler
import android.os.Looper

/**
 * Streams OS-validated network status to Flutter.
 *
 * Uses [ConnectivityManager.registerDefaultNetworkCallback]; the key signal is
 * [NetworkCapabilities.NET_CAPABILITY_VALIDATED] - the system's own
 * internet/captive-portal validation - which the recovery pipeline uses to
 * tell "no internet" from "device unreachable".
 */
class NetworkMonitorApiImpl(
  context: Context,
  private val events: NetworkMonitorEvents,
) : NetworkMonitorApi {
  private val connectivityManager =
    context.applicationContext.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
  private val mainHandler = Handler(Looper.getMainLooper())
  private var networkCallback: ConnectivityManager.NetworkCallback? = null

  override fun getCurrentStatus(): NativeNetworkStatus =
    statusFrom(connectivityManager.getNetworkCapabilities(connectivityManager.activeNetwork))

  override fun startObserving() {
    if (networkCallback != null) {
      return
    }
    val callback = object : ConnectivityManager.NetworkCallback() {
      override fun onCapabilitiesChanged(network: Network, capabilities: NetworkCapabilities) {
        emit(statusFrom(capabilities))
      }

      override fun onLost(network: Network) {
        emit(statusFrom(null))
      }
    }
    networkCallback = callback
    connectivityManager.registerDefaultNetworkCallback(callback)
  }

  override fun stopObserving() {
    networkCallback?.let { connectivityManager.unregisterNetworkCallback(it) }
    networkCallback = null
  }

  private fun emit(status: NativeNetworkStatus) {
    // Platform channels must be used from the main thread; connectivity
    // callbacks arrive on a ConnectivityThread handler.
    mainHandler.post {
      events.onStatusChanged(status) {}
    }
  }

  private fun statusFrom(capabilities: NetworkCapabilities?): NativeNetworkStatus {
    if (capabilities == null) {
      return NativeNetworkStatus(
        hasTransport = false,
        transports = emptyList(),
        internetValidated = false,
        isExpensive = false,
      )
    }

    val transports = buildList {
      if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) ||
        capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI_AWARE)
      ) {
        add(NativeTransportType.WIFI)
      }
      if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR)) {
        add(NativeTransportType.CELLULAR)
      }
      if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET)) {
        add(NativeTransportType.ETHERNET)
      }
      if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_VPN)) {
        add(NativeTransportType.VPN)
      }
    }

    return NativeNetworkStatus(
      hasTransport = true,
      transports = transports.ifEmpty { listOf(NativeTransportType.OTHER) },
      internetValidated = capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED),
      isExpensive = !capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_NOT_METERED),
    )
  }
}
